### PR TITLE
#906 check limits during pin tip centring

### LINKED ISSUE #906: move_pin_into_view moves out of smargon x limits
when the pin is too long 'move_pin_into_view' moves out of the smargon x limit during OAV centring

https://ispyb.diamond.ac.uk/dc/visit/mx32736-17/id/11772912
https://graylog2.diamond.ac.uk/messages/logs_8415/81e663d1-5d84-11ee-ae6d-1866dafadc6c

we should check the limits and raise a warningexception to exit the experiment gracefully

### PR DESCRIPTION
**Pull request by dperl-dls**
- move to the limit if normal nudge would take it over since into-view steps are large
- update tests with limits

Fixes #906

### To test:
1. Run tests

### CODE REVIEW

File: src/hyperion/parameters/constants.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 # this one is for making depositions:
 DEV_ISPYB_DATABASE_CFG = "/dls_sw/dasc/mariadb/credentials/ispyb-dev.cfg"
 PARAMETER_SCHEMA_DIRECTORY = "src/hyperion/parameters/schemas/"
+OAV_REFRESH_DELAY = 0.3
```
Comment (DominicOram) [inline src/hyperion/parameters/constants.py:17]: Nit: I feel like this should be in the dodal device?

File: src/hyperion/experiment_plans/pin_tip_centring_plan.py [hunk 2]
```diff
@@ -49,16 +56,34 @@ def move_pin_into_view(
...
-            LOGGER.warning(f"Pin is too long, moving -{step_size}mm")
-            yield from bps.mvr(smargon.x, -step_size)
+            min_limit = min(smargon.x.limits)
+            if float(smargon.x.user_readback.get()) - step_size_mm < min_limit:
+                LOGGER.warning(
+                    f"Pin is too long, moving -{step_size_mm} mm would cross limits, "
+                    f"moving to minimum limit {min_limit}"
+                )
+                yield from bps.mv(smargon.x, min_limit)
+                yield from bps.sleep(OAV_REFRESH_DELAY)
```
Comment (DominicOram) [inline src/hyperion/experiment_plans/pin_tip_centring_plan.py:66]: Should: Given we're doing this sleep again below does this mean it sleeps twice? I don't think that's necessary

File: src/hyperion/experiment_plans/pin_tip_centring_plan.py [hunk 3]
```diff
@@ -49,16 +56,34 @@ def move_pin_into_view(
         tip_x_px, tip_y_px = yield from bps.rd(oav.mxsc.pin_tip)
 
         if tip_x_px == 0:
-            LOGGER.warning(f"Pin is too long, moving -{step_size}mm")
-            yield from bps.mvr(smargon.x, -step_size)
+            min_limit = min(smargon.x.limits)
+            if float(smargon.x.user_readback.get()) - step_size_mm < min_limit:
```
Comment (DominicOram) [inline src/hyperion/experiment_plans/pin_tip_centring_plan.py:60]: Must: We should be using `rd` in the plan, not `get`

File: src/hyperion/experiment_plans/pin_tip_centring_plan.py [hunk 4]
```diff
@@ -49,16 +56,34 @@ def move_pin_into_view(
...
+                yield from bps.mv(smargon.x, min_limit)
+                yield from bps.sleep(OAV_REFRESH_DELAY)
+                break
+            LOGGER.warning(f"Pin is too long, moving -{step_size_mm}mm")
+            yield from bps.mvr(smargon.x, -step_size_mm)
         elif tip_x_px == oav.mxsc.pin_tip.INVALID_POSITION[0]:
-            LOGGER.warning(f"Pin is too short, moving {step_size}mm")
-            yield from bps.mvr(smargon.x, step_size)
+            max_limit = max(smargon.x.limits)
+            if float(smargon.x.user_readback.get()) - step_size_mm > max_limit:
```
Comment (DominicOram) [inline src/hyperion/experiment_plans/pin_tip_centring_plan.py:72]: Must: This should be a `+` not a `-`

**Review by DominicOram — CHANGES_REQUESTED** [review]
Comments in code. Also, as a could, I feel like we might be able to make this cleaner by doing something like:
* Invert step size if too long (not too short)
* Calculate ideal absolute move ourselves
* Clip to limits
* Do the absolute move

File: src/hyperion/experiment_plans/pin_tip_centring_plan.py (hunk 4 above)
Comment (DominicOram) [reply src/hyperion/experiment_plans/pin_tip_centring_plan.py:72]: Must: Also, can we get a test to cover this?

File: src/hyperion/experiment_plans/pin_tip_centring_plan.py (hunk 2 above)
Comment (dperl-dls) [reply src/hyperion/experiment_plans/pin_tip_centring_plan.py:66]: no, break gets out of the for loop so we won't hit the other sleep

Comment (DominicOram) [reply src/hyperion/experiment_plans/pin_tip_centring_plan.py:66]: Yep, sorry, my bad. I think this leads credence to the general comment of finding a slightly nicer way of doing it then. Given that we're repeating this line and the logic isn't 100% obvious

**Review by DominicOram — APPROVED** [review]
Looks good, thank you. I still think it could be a bit cleaner but it's a nit pick

**Comment by dperl-dls** [comment]
I agree it could be cleaner, and I'll be happy if someone improves it, but it works correctly and I don't have time right now

**Pull request merged**